### DISCUSSION TITLE
Use msgpack to serialize/deserialize the payload

### DISCUSSION THREAD

**Discussion by mzy2240** [writer/writer-framework#9] [Ideas]
In most cases, msgpack could help reduce the communication latency. I have very good experience using ormsgpack in the backend and msgpackr in the frontend.

**Comment by ramedina86** [comment]
This is something I want to explore.

I'm confident serialisation in the backend would be fast and comparable to orjson, but I have my doubts when it comes to client-side deserialisation. Given that JSON parsing is part of the V8 engine and highly optimised, I've seen benchmarks of JSON outperforming Protocol Buffers. Plus, the added complexity. Is this something you'd be interested in testing downstream first?

Also, for dataframes specifically (which I expect will constitute 90% of state mutation payloads) I'll look into Pyarrow. For dataframes, JSON works but seems like a simplistic approach and it'd make it trickier to support non-pandas dataframes.

**Comment by mzy2240** [comment]
Yeah I tested in my own fork before, and I can ensure you the performance of ormsgpack + msgpackr is incredible. The benefit of using msgpack, besides its performance, also includes the capability to keep raw format (like list, tuple, array, etc.), instead of the hassle to wrap everything into a JSON. Sometimes converting to dictionary (and JSON) brings extra cost.

**Comment by mzy2240** [comment]
Arrow is a good idea for dataframe, it also enables you to handle dataframe from Polars in the future
